### PR TITLE
Change missing or empty secret error message

### LINKED ISSUE #2023: Change error message in retrieve variable if variable is empty
As a developer, I want to understand the different reasons for failing to retrieve a Conjur variable  , so that I can identify the problem better and resolve the it if needed.

Today when one retrieve a variable and the variable is empty he gets 
Variable <x> not found in account <y>
We would like to change only the message to
Variable <x> is empty or not found

GIVEN policy with variable X loaded
WHEN retrieving variable x
THEN return 404 with error message Variable <x> is empty or not found

## Developer Notes
- http code should remain the same
- code error in the message should remain the same

### PR DESCRIPTION
**Pull request by liavyona**
### What does this PR do?
A new error message while fetching missing or empty secret

### What ticket does this PR close?
Resolves #2023 , #2083 

#### Change log
- [x] The CHANGELOG has been updated, or
- [ ] This PR does not include user-facing changes and doesn't require a CHANGELOG update

#### Test coverage
- [x] This PR includes new unit and integration tests to go with the code changes, or
- [ ] The changes in this PR do not require tests

#### Documentation
- [ ] Docs (e.g. `README`s) were updated in this PR, and/or there is a follow-on issue to update docs, or
- [x] This PR does not require updating any documentation

#### API Changes
- [ ] The [OpenAPI spec](https://github.com/cyberark/conjur-openapi-spec) has been updated to meet new API changes (or an issue has been opened), or
- [ ] The changes in this PR do not affect the Conjur API

image

### CODE REVIEW

**Review by orenbm — DISMISSED** [review]
LGTM! please let @shulifink review this as well

File: app/domain/errors.rb [hunk 1]
```diff
@@ -31,6 +31,11 @@ module Conjur
       code: "CONJ00074E"
     )
 
+    MissingSecretValue = ::Util::TrackableErrorClass.new(
+      msg:  "Variable {0-variable-id} is empty or not found.",
```
Comment (liavyona) [inline app/domain/errors.rb:35]: @shulifink  Is this error approved? @InbalZilberman @orenbm Don't we want the account?

Comment (InbalZilberman) [reply app/domain/errors.rb:35]: @liavyona @shulifink confirmed she is fine with the message in the issue, if you remember. fire away!

**Review by InbalZilberman — DISMISSED** [review]
LGTM! 10X

File: CHANGELOG.md [hunk 2]
```diff
@@ -15,6 +15,8 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 ### Changed
 - The secrets batch retrieval endpoint now refers to the `Accept-Encoding` header rather than `Accept` to determine the response encoding
   [cyberark/conjur#2065](https://github.com/cyberark/conjur/pull/2065)
+- Return a proper error message when trying to fetch a missing or empty secret
```
Comment (izgeri) [inline CHANGELOG.md:18]:
```suggestion
- Return a proper error message when trying to fetch a missing or empty secret.
```

Comment (shulifink) [reply CHANGELOG.md:18]:
```suggestion
- When trying to fetch a missing or empty secret, a proper error message is returned
```

File: app/domain/errors.rb (hunk 1 above)
Comment (shulifink) [reply app/domain/errors.rb:35]: fine

**Review by shulifink — CHANGES_REQUESTED** [review]
small change - remember, in the Change log, we want to tell the reader what changed.
So instead of saying "_Do this..._."  we would want to say " _We changed this...._"

File: cucumber/api/features/secrets_batch.feature [hunk 3]
```diff
@@ -54,19 +54,27 @@ Feature: Batch retrieval of secrets
...
     Then the HTTP response status code is 404
+    And there is an error
+    And the error message includes "is empty or not found"
 
   Scenario: Fails with 404 if a resource doesn't have a value
     Given I create a new "variable" resource called "secret-no-value"
     When I GET "/secrets?variable_ids=cucumber:variable:secret1,cucumber:variable:secret-no-value"
     Then the HTTP response status code is 404
+    And there is an error
+    And the error message includes "is empty or not found"
```
Comment (orenbm) [inline cucumber/api/features/secrets_batch.feature:77]: do you want to verify that the account is written as well? you can verify that the message includes `Variable cucumber:variable:secret-no-value is empty or not found`

File: cucumber/api/features/secrets_batch.feature [hunk 4]
```diff
@@ -53,20 +53,28 @@ Feature: Batch retrieval of secrets
 
   Scenario: Fails with 404 if variable_ids param has some blank items
     When I GET "/secrets?variable_ids=cucumber:variable:secret1,,,cucumber:variable:secret2"
-    Then the HTTP response status code is 404
+    Then the HTTP response status code is 422
```
Comment (orenbm) [inline cucumber/api/features/secrets_batch.feature:56]: looks like you are breaking the API. did @netacoral approve this? i'm ok with leaving it on 404.

Comment (liavyona) [reply cucumber/api/features/secrets_batch.feature:56]: It is the same as only blank requests. @InbalZilberman @netacoral What do you say?

Comment (orenbm) [reply cucumber/api/features/secrets_batch.feature:56]: looping in @izgeri as well

Comment (liavyona) [reply cucumber/api/features/secrets_batch.feature:56]: @izgeri @orenbm I added verification that batch request doesn't contain nameless secrets ('/?variable_ids=var1,,,,,,var2'). Previously, such a request got a 404 response. In the test file `secrets_batch.feature` you can find that a request with '/?variable_ids=,,,,,,' returns `422`. The verification change caused this API change. Is that OK?

Comment (InbalZilberman) [reply cucumber/api/features/secrets_batch.feature:56]: @orenbm thank you pointing this out and reaching out.
If you look few tests above when providing ",,," as var ids we give 422 as our response code, hence this change is not breaking behaviour. The fact that this test provided 404 is actually a mystery from me. 
I think that returning 422 here is the right action and actually the previous test is wrong.
WDYT?

**Review by shulifink — APPROVED** [review]

**Review by orenbm — APPROVED** [review]

**Review by diverdane — COMMENTED** [review]
A couple of minor editorial suggestions. There are still some CodeClimate errors that need to be addressed (https://codeclimate.com/github/cyberark/conjur/pull/2074).  I'll defer to @izgeri re. whether this is an okay API change, but it seems reasonable to me.

**Comment by liavyona** [comment]
@diverdane Codeclimate steps are related to application_controller formatting. All functions there gets e as parameter and use the json formatting for response.... This is code conventions

File: CHANGELOG.md [hunk 5]
```diff
@@ -17,6 +21,8 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 ### Changed
 - The secrets batch retrieval endpoint now refers to the `Accept-Encoding` header rather than `Accept` to determine the response encoding
   [cyberark/conjur#2065](https://github.com/cyberark/conjur/pull/2065)
+- When trying to fetch a missing or empty secret, a proper error message is now returned
```
[the remark this answers is not in the document]
Comment (izgeri) [reply CHANGELOG.md:24]:
```suggestion
- When trying to fetch a missing or empty secret, a proper error message is now returned.
```

**Review by diverdane — APPROVED** [review]
@liavyona Thanks for the explanation on the CodeClimate issues! I agree, it looks like the CodeClimate CI "errors" in this case can be considered noise. I chatted with some folks here re. the more general question of how we should handle CodeClimate issues in community repos, and we're trying to fine-tune the guidelines.

So, aside from the couple of editorial nits, it LGTM!

**Review by tzheleznyak — APPROVED** [review]
LGTM

**Pull request merged**